### PR TITLE
Implement NTP Service Montor

### PR DESCRIPTION
**Pull request by Mohammad-Junid**
## Description
Build a NTP service monitor and register it in minion.

## Jira link(s)
- https://opennms.atlassian.net/browse/LOK-2365

## Checklist
* [ ] Follows Lōkahi's [development guidelines.](https://github.com/OpenNMS-Cloud/lokahi/wiki/Development-Guidelines)
* [ ] Appropriate reviewer(s) have been selected.
* [ ] Jira issue(s) have been updated to "In Review".
* [ ] Includes [appropriate tests.](https://github.com/OpenNMS-Cloud/lokahi/wiki/Test-Strategy)
* [ ] Documentation has been updated as necessary.
* [ ] Notify devops of changes to the Charts.
* [ ] Notify documentation team of any changes to names of screens or features (affects URLs).

### CODE REVIEW

File: minion/plugins/ntp-plugin/pom.xml [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+  <parent>
+    <groupId>org.opennms.lokahi.minion</groupId>
+    <artifactId>plugins</artifactId>
+    <version>0.1.0-SNAPSHOT</version>
+  </parent>
+
+  <artifactId>ntp-plugin</artifactId>
+  <packaging>bundle</packaging>
+
+  <name>OpenNMS Lokahi :: Minion :: Plugins :: HTTP</name>
```
Comment (smunir-onms) [inline minion/plugins/ntp-plugin/pom.xml:34]: <name>OpenNMS Lokahi :: Minion :: Plugins :: NTP</name>

File: minion/plugins/ntp-plugin/pom.xml [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+                <descriptor>mvn:org.apache.karaf.features/framework/${karaf.version}/xml/features</descriptor>
+                <descriptor>mvn:org.apache.karaf.features/standard/${karaf.version}/xml/features</descriptor>
+                <descriptor>mvn:org.opennms.lokahi.minion/grpc-features/${project.version}/xml/features</descriptor>
+                <descriptor>mvn:org.opennms.lokahi.minion/ipc-features/${project.version}/xml/features</descriptor>
+                <descriptor>mvn:org.opennms.lokahi.minion/observability-features/${project.version}/xml/features</descriptor>
+                <descriptor>mvn:org.opennms.lokahi.minion/plugin-features/${project.version}/xml/features</descriptor>
+                <descriptor>file:${project.build.directory}/feature/feature.xml</descriptor>
+              </descriptors>
+              <features>
+                <feature>icmp*</feature>
```
[the remark this answers is not in the document]
Comment (Mohammad-Junid) [reply minion/plugins/ntp-plugin/pom.xml:135]: Done

File: minion/plugins/ntp-plugin/pom.xml (hunk 1 above)
Comment (Mohammad-Junid) [reply minion/plugins/ntp-plugin/pom.xml:34]: Done

**Review by cgorantla — COMMENTED** [review]
Please fix Minion build failure

File: minion/plugins/ntp-plugin/pom.xml [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+      <groupId>junit</groupId>
+      <artifactId>junit</artifactId>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.mockito</groupId>
+      <artifactId>mockito-core</artifactId>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
```
Comment (cgorantla) [inline minion/plugins/ntp-plugin/pom.xml:71]: Do we need event conf here ?

File: minion/plugins/ntp-plugin/src/main/java/org/opennms/lokahi/minion/ntp/NtpMonitorManager.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+import lombok.RequiredArgsConstructor;
+import org.opennms.horizon.minion.plugin.api.ServiceMonitor;
+import org.opennms.horizon.minion.plugin.api.ServiceMonitorManager;
+
+@RequiredArgsConstructor
+public class NtpMonitorManager implements ServiceMonitorManager {
+
+    @Override
+    public ServiceMonitor create() {
+        NtpMonitor httpMonitor = new NtpMonitor();
```
Comment (cgorantla) [inline minion/plugins/ntp-plugin/src/main/java/org/opennms/lokahi/minion/ntp/NtpMonitorManager.java:33]:
```suggestion
        NtpMonitor ntpMonitor = new NtpMonitor();
```

File: minion/plugins/ntp-plugin/src/main/resources/OSGI-INF/blueprint/ntp-plugins.xml [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+           xmlns:camel="http://camel.apache.org/schema/blueprint"
+           xmlns="http://www.osgi.org/xmlns/blueprint/v1.0.0"
+           xsi:schemaLocation="http://www.osgi.org/xmlns/blueprint/v1.0.0 http://www.osgi.org/xmlns/blueprint/v1.0.0/blueprint.xsd">
+
+    <cm:property-placeholder persistent-id="org.opennms.minion.plugins.ntp" update-strategy="reload">
+        <cm:default-properties>
+        </cm:default-properties>
+    </cm:property-placeholder>
+
+    <bean id="httpMonitorManager" class="org.opennms.lokahi.minion.ntp.NtpMonitorManager"/>
```
Comment (cgorantla) [inline minion/plugins/ntp-plugin/src/main/resources/OSGI-INF/blueprint/ntp-plugins.xml:13]:
```suggestion
    <bean id="ntpMonitorManager" class="org.opennms.lokahi.minion.ntp.NtpMonitorManager"/>
```

**Review by cgorantla — COMMENTED** [review]
looks good, overall.

Can you validate with karaf command and post results here.

File: minion/plugins/ntp-plugin/src/main/feature/feature.xml [hunk 6]
```diff
@@ -0,0 +1,19 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<features name="minion-plugin-ntp-${project.version}" xmlns="http://karaf.apache.org/xmlns/features/v1.6.0">
+
+
+    <feature name="ntp-shared-libs" hidden="true">
+        <feature>aries-blueprint</feature>
+        <feature>grpc-contract</feature>
+        <bundle dependency="true">mvn:org.opennms.lokahi.shared/lokahi-common-logging/${project.version}</bundle>
+        <bundle dependency="true">mvn:org.opennms.lokahi.shared/lokahi-common-utils/${project.version}</bundle>
+        <bundle dependency="true">mvn:org.opennms.lib/org.opennms.lib.tracker/0.7</bundle>
```
Comment (cgorantla) [inline minion/plugins/ntp-plugin/src/main/feature/feature.xml:10]: this bundle is not needed

File: minion/plugins/ntp-plugin/pom.xml (hunk 3 above)
Comment (Mohammad-Junid) [reply minion/plugins/ntp-plugin/pom.xml:71]: Did need the event conf I removed it

File: minion/plugins/ntp-plugin/src/main/java/org/opennms/lokahi/minion/ntp/NtpMonitorManager.java (hunk 4 above)
Comment (Mohammad-Junid) [reply minion/plugins/ntp-plugin/src/main/java/org/opennms/lokahi/minion/ntp/NtpMonitorManager.java:33]: Change instance name

File: minion/plugins/ntp-plugin/src/main/resources/OSGI-INF/blueprint/ntp-plugins.xml (hunk 5 above)
Comment (Mohammad-Junid) [reply minion/plugins/ntp-plugin/src/main/resources/OSGI-INF/blueprint/ntp-plugins.xml:13]: bean id changed

File: minion/plugins/ntp-plugin/src/main/feature/feature.xml (hunk 6 above)
Comment (Mohammad-Junid) [reply minion/plugins/ntp-plugin/src/main/feature/feature.xml:10]: removed useless bundles

Comment (Mohammad-Junid) [reply minion/plugins/ntp-plugin/src/main/feature/feature.xml:10]: [Screencast from 15-04-2024 22:36:10.webm]

File: shared-lib/protobuf/src/main/proto/ntp.proto [hunk 7]
```diff
@@ -0,0 +1,18 @@
+syntax = "proto3";
+
+import "google/protobuf/any.proto";
+
+
+package opennms.icmp;
```
Comment (smunir-onms) [inline shared-lib/protobuf/src/main/proto/ntp.proto:6]: package opennms.ntp;

Comment (Mohammad-Junid) [reply shared-lib/protobuf/src/main/proto/ntp.proto]: Done

File: minion/plugins/ntp-plugin/src/main/feature/feature.xml [hunk 8]
```diff
@@ -0,0 +1,19 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<features name="minion-plugin-ntp-${project.version}" xmlns="http://karaf.apache.org/xmlns/features/v1.6.0">
+
+
+    <feature name="ntp-shared-libs" hidden="true">
+        <feature>aries-blueprint</feature>
+        <feature>grpc-contract</feature>
```
Comment (cgorantla) [inline minion/plugins/ntp-plugin/src/main/feature/feature.xml:7]: don't these we need these two as well

**Review by cgorantla — COMMENTED** [review]
This looks good. We may need to add Inventory info to the response once https://github.com/OpenNMS-Cloud/lokahi/pull/1758 is taken care.

File: minion/plugins/ntp-plugin/src/main/feature/feature.xml (hunk 8 above)
Comment (Mohammad-Junid) [reply minion/plugins/ntp-plugin/src/main/feature/feature.xml:7]: I don't think it is needed I removed it

**Review by cgorantla — APPROVED** [review]
LGTM!

**Pull request merged**